machinist/mserver/controller.go

```diff
@@ -7,11 +7,22 @@ import (
 )
 
 type Controller struct {
-	workers        sync.Map
-	ConnectedNodes map[string][]string
+	connectedNodes map[string]*Node
```

Comment:
Looks better, thanks. Now... related to the comment below, about public methods:
- by embedding sync.RWMutex directly, anyone can call `Controller.Lock()` and `Controller.Unlock`. Is that desired? Generally public interfaces of an object should not require the **caller** to know how to perform locking on the **callee** (makes changing and optimizing locking later harder, makes all the callsites potentially causing deadlocks/mishandling)
- further, as the code grows, I generally end up with some parts of an object *not* protected by lock (eg, read only, set at initialization) and other parts read write, or under different locks.

My recommendation would be to change to:
```
type Controller struct {
  connectedNodes map[string]*Node // All access protected by c_lock below.
  c_lock sync.RWMutex
}
```

(eg, name locks, so they are not accessible from outside, add comments so it's clear what's locked by what)

Reply:
looking how controlledNodes is used so far, I actually recommend having the locking directly in Nodes(), Node(), and adding a "Set()" method (also doing locking). `c_lock` would then only be used by those few methods.

(I would also consider creating a dedicated NodeSet data structure, for example, so the public Controller methods don't get muddied with the private NodeSet methods, and you can *guarantee* and *enforce* using the locks on NodeSet, via its public methods - any Controller method can otherwise access c_lock directly) (entirely optional now, as Controller, well, there's not much there anyway).